src/validation.cpp

```diff
@@ -620,19 +621,70 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
             std::string dummy_err_string;
-            // If the new transaction is relatively small (up to 40k weight)
-            // and has at most one ancestor (ie ancestor limit of 2, including
-            // the new transaction), allow it if its parent has exactly the
-            // descendant limit descendants.
-            //
-            // This allows protocols which rely on distrusting counterparties
-            // being able to broadcast descendants of an unconfirmed transaction
-            // to be secure by simply only having two immediately-spendable
-            // outputs - one for each counterparty. For more info on the uses for
-            // this, see https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-November/016518.html
-            if (nSize >  EXTRA_DESCENDANT_TX_SIZE_LIMIT ||
-                    !pool.CalculateMemPoolAncestors(entry, setAncestors, 2, nLimitAncestorSize, nLimitDescendants + 1, nLimitDescendantSize + EXTRA_DESCENDANT_TX_SIZE_LIMIT, dummy_err_string)) {
-                return state.Invalid(ValidationInvalidReason::TX_MEMPOOL_POLICY, false, REJECT_NONSTANDARD, "too-long-mempool-chain", errString);
+            if (setConflicts.size() == 1) {
+                // In general, when we receive an RBF transaction with mempool conflicts, we want to know whether we
+                // would meet the chain limits after the conflicts have been removed. However, there isn't a practical
+                // way to do this short of calculating the ancestor and descendant sets with an overlay cache of
+                // changed mempool entries. Due to both implementation and runtime complexity concerns, this isn't
+                // very realistic, either, thus we we try to ensure a limited set of transaction is RBF'able despite
+                // mempool conflicts here. Importantly, we want to ensure that transactions which were accepted using
+                // the below carve-out are able to be RBF'ed, without impacting the security the carve-out provides
+                // for off-chain contract systems (see link in the comment below).
+                //
+                // Specifically, the subset of RBF transactions which we allow despite conflicts are those which
+                // conflict directly with exactly one other transaction (but may evict children of said transaction),
+                // and which are not adding any new mempool dependencies (checked first).
+                // Such transactions are clearly not merging any existing packages, so we are only concerned with
+                // ensuring that (a) no package is growing past the package size (not count) limits and (b) we are
+                // not allowing something to effectively use the (below) carve-out spot when it shouldn't be allowed
+                // to.
+                // To check these we re-run CalculateMemPoolAncestors up to twice, once corresponding to the above,
```

Comment:
Having real trouble parsing this last paragraph.

Lots of different things being referenced, "above", "below", "corresponding", unsure exactly what it's referring to.

Reply:
Is this a bit more readable given the changes made for https://github.com/bitcoin/bitcoin/pull/16421#pullrequestreview-275251295 ?